R/splitRaster.R

```diff
@@ -80,69 +92,139 @@ setGeneric("splitRaster", function(x, nx, ny, buffer) {
 setMethod(
   "splitRaster",
   signature = signature(x = "RasterLayer", nx = "integer",
-                        ny = "integer", buffer = "numeric"),
-  definition = function(x, nx, ny, buffer) {
+                        ny = "integer", buffer = "numeric",
+                        savePath = "character"),
+  definition = function(x, nx, ny, buffer, savePath) {
     if (length(buffer) > 2) {
       warning("buffer contains more than 2 elements - only the first two will be used.")
       buffer <- buffer[1:2]
     } else if (length(buffer) == 1) {
       buffer <- c(buffer, buffer)
     }
-
     if (buffer[1] < 1) {
       buffer[1] <- ceiling((buffer[1]*(xmax(x) - xmin(x))/nx)/xres(x))
     }
     if (buffer[2] < 1) {
       buffer[2] <- ceiling((buffer[2]*(ymax(x) - ymin(x))/ny)/yres(x))
     }
     ext <- extent(x)
-    tiles <- vector("list", length = nx*ny)
+    extents <- vector("list", length = nx*ny)
+    if(!dir.exists(savePath)) {dir.create(savePath)}
     n <- 1L
     for (i in seq_len(nx) - 1L) {
       for (j in seq_len(ny) - 1L) {
         x0 <- ext@xmin + i*((ext@xmax - ext@xmin) / nx) - buffer[1]*xres(x)
         x1 <- ext@xmin + (i + 1L)*((ext@xmax - ext@xmin) / nx) + buffer[1]*xres(x)
         y0 <- ext@ymin + j*((ext@ymax - ext@ymin) / ny) - buffer[2]*yres(x)
         y1 <- ext@ymin + (j + 1L)*((ext@ymax - ext@ymin) / ny) + buffer[2]*yres(x)
-        tiles[[n]] <- crop(x, extent(x0, x1, y0, y1))
+        extents[[n]] <- extent(x0, x1, y0, y1)
         n <- n + 1L
       }
     }
+    cl <- tryCatch(getCluster(), error = function(x) NULL)
+    on.exit(if (!is.null(cl)) returnCluster())
+    if (!is.null(cl)) {
+      parFun <- "clusterApply"
+      cropNew <- function(ras, x0 = x, ...) {
+        crop(x = ras,y = x0)
+      }
+      args <- list(cl = cl, x = extents,
+                   fun = cropNew, ras = x)
+
+      # args <- list(seq = extents, fun = crop, r)
+      # args1 <- list(cl = cl, y = extents, fun = crop, x = r)
+      # args <- append(list(cl = cl), args)
+    } else {
+      parFun <- "lapply"
+      args <- list(extents, FUN = crop, x = r)
+    }
+    tiles <- do.call(get(parFun), args)
+    for(i in 1:length(tiles)){
```

Comment:
alex,

could you please provide the repeatable example, so that I can debug codes?
I have tested the function using below codes, and did not find something
wrong.

library(raster)
  on.exit(detach("package:raster"))

  b <- brick(system.file("external/rlogo.grd", package = "raster"))
  r <- b[[1]] # use first layer only
  nx <- 3
  ny <- 4
  expect_equal(xres(r), 1)
  expect_equal(yres(r), 1)

  # change the extent of r
  extent(r) <- extent(xmin(r) - 30, xmax(r) - 30, ymin(r) - 20, ymax(r) -
20)
  orgwd <- getwd()

  y1 <- splitRaster(r, nx, ny, c(3L, 4L), savePath = file.path(tempdir(),
"red1"))
  expect_true(unique(unlist(lapply(y1, fromDisk))))

  for(i in 1:12){
    expect_true(file.exists(file.path(getwd(), "red1",
                                      paste("red_tile", i, ".tif", sep =
""))))
  }

  xextents <- c()
  yextents <- c()
  for (i in 1:length(y0)) {
    xextents <- c(xextents, xmin(y1[[i]]), xmax(y1[[i]]))
    yextents <- c(yextents, ymin(y1[[i]]), ymax(y1[[i]]))
  }
  expect_equal(sort(unique(xextents)), c(-30, 1, 7, 34, 40, 71))
  expect_equal(sort(unique(yextents)), c(-20, -5, 3, 14, 22, 34, 42, 57))
  rm(xextents, yextents)

  expect_equal(length(unique(lapply(y1, crs))), 1L)
  expect_equal(unique(lapply(y1, crs))[[1]], crs(r))

  m1 <- mergeRaster(y1)
  expect_equal(dim(m1), dim(r))
  expect_equal(extent(m1), extent(r))
  expect_equal(names(m1), names(r)) # TO DO (#283)
  expect_equal(res(m1), res(r))
  expect_equal(max(values(m1)), max(values(r))) # TO DO (#283)
  expect_equal(min(values(m1)), min(values(r))) # TO DO (#283)
  unlink(file.path(tempdir(), "red1"), recursive = TRUE)

  #test parallel cropping
  beginCluster(2)
  cl <- getCluster()
  expect_equal(length(cl), 2)
  y11 <- splitRaster(r, nx, ny, c(3L, 4L), savePath = file.path(tempdir(),
"red11"))
  expect_true(unique(unlist(lapply(y1, fromDisk))))

  for(i in 1:12){
    expect_true(file.exists(file.path(getwd(), "red11",
                                      paste("red_tile", i, ".tif", sep =
""))))
  }

  xextents <- c()
  yextents <- c()
  for (i in 1:length(y11)) {
    xextents <- c(xextents, xmin(y11[[i]]), xmax(y11[[i]]))
    yextents <- c(yextents, ymin(y11[[i]]), ymax(y11[[i]]))
  }
  expect_equal(sort(unique(xextents)), c(-30, 1, 7, 34, 40, 71))
  expect_equal(sort(unique(yextents)), c(-20, -5, 3, 14, 22, 34, 42, 57))
  rm(xextents, yextents)

  expect_equal(length(unique(lapply(y11, crs))), 1L)
  expect_equal(unique(lapply(y11, crs))[[1]], crs(r))

  m11 <- mergeRaster(y11)
  expect_equal(dim(m11), dim(r))
  expect_equal(extent(m11), extent(r))
  expect_equal(names(m11), names(r)) # TO DO (#283)
  expect_equal(res(m11), res(r))
  expect_equal(max(values(m11)), max(values(r))) # TO DO (#283)
  expect_equal(min(values(m11)), min(values(r))) # TO DO (#283)
  unlink(file.path(tempdir(), "red11"), recursive = TRUE)
  endCluster()

for(i in 1:12){ print(extent(y1[[i]])==extent(y11[[i]]))} # all true


On Thu, Jun 16, 2016 at 3:23 PM, Alex Chubaty <notifications@github.com>
wrote:

> In R/splitRaster.R
> <https://github.com/PredictiveEcology/SpaDES/pull/286#discussion_r67435044>
> :
>
> > +      parFun <- "clusterApply"
> > +      cropNew <- function(ras, x0 = x, ...) {
> > +        crop(x = ras,y = x0)
> > +      }
> > +      args <- list(cl = cl, x = extents,
> > +                   fun = cropNew, ras = x)
> > +
> > +      # args <- list(seq = extents, fun = crop, r)
> > +      # args1 <- list(cl = cl, y = extents, fun = crop, x = r)
> > +      # args <- append(list(cl = cl), args)
> > +    } else {
> > +      parFun <- "lapply"
> > +      args <- list(extents, FUN = crop, x = r)
> > +    }
> > +    tiles <- do.call(get(parFun), args)
> > +    for(i in 1:length(tiles)){
>
> something here isn't right -- the resulting raster tiles are messed up
> when not using a cluster.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/PredictiveEcology/SpaDES/pull/286/files/a733cb411ca2c41632018cd6f2372a7bd13146e0#r67435044>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AJ-1fLlks1Et0J4FHELL3M83JrLaQBCWks5qMcz6gaJpZM4I3FuJ>
> .
>



-- 
Yong Luo
Postdoctoral Researcher

Canadian Forest Service
Natural Resources Canada
Pacific Forestry Centre
506 Burnside Road W
Victoria, BC  V8Z 1M5
Canada
phone: +1-250-298-2566
&
Faculty of Forestry
University of British Columbia
2424 Main Mall
Vancouver, BC V6T 1Z4
Canada